### PR TITLE
add filesystem cache

### PR DESCRIPTION
**Pull request by chungwong**
This is my attempt at https://github.com/GoogleChrome/rendertron/issues/338

### CODE REVIEW

File: src/filesystem-cache.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  async clearAllCache() {
+    fs.rmdirSync(this.getDir(''), {recursive: true});
+  }
+
+  cacheContent(key: string, ctx: Koa.Context) {
+    const response = ctx.response;
+    const payload = ctx.body;
+    const request = ctx.request;
```
Comment (AVGP) [inline src/filesystem-cache.ts:81]: What are we storing the request for?

File: src/filesystem-cache.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  async clearCache(key: string) {
+    fs.rmdirSync(this.getDir(key), {recursive: true});
+  }
+
+  async clearAllCache() {
+    fs.rmdirSync(this.getDir(''), {recursive: true});
+  }
+
+  cacheContent(key: string, ctx: Koa.Context) {
+    const response = ctx.response;
```
Comment (AVGP) [inline src/filesystem-cache.ts:79]: To make naming a bit more clear, what do you think about naming this variable `responseHeaders` or `responseMeta` or something similar? Then the `payload` could either stay with that name or be renamed `responseBody` to make the distinction clearer.

File: src/filesystem-cache.ts [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+      }
+
+      // remove /render/ from key
+      cacheKey = cacheKey.replace(/^\/render\//, '');
+
+      // remove trailing slash from key
+      cacheKey = cacheKey.replace(/\/$/, '');
+
+      // key is hashed crudely
+      const key = hashCode(cacheKey);
```
[the remark this answers is not in the document]
Comment (chungwong) [reply src/filesystem-cache.ts:149]: in my mind I was thinking using md5 wouldl generated 32 chars which might be a bit too long.
If 32 chars would not caused any trouble, I can definitely change to use MD5

File: src/filesystem-cache.ts (hunk 1 above)
Comment (chungwong) [reply src/filesystem-cache.ts:81]: in my use case, if I want to question about the cached generated, I would like to know what the actual request(url, user-agent) was about, the current design for storing cache didn't store the request URL if I read it correctly and that was why I decided to save the request too.

File: src/filesystem-cache.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (AVGP) [reply src/filesystem-cache.ts:149]: I think that would be good.

**Comment by chungwong** [comment]
I have updated this PR, I am leaving out the serialisation of headers, body and request and putting all of them in a single file as I am actually have to work on something else ASAP.

**Comment by AVGP** [comment]
@chungwong Hey thanks again for your contribution! It's not ready to be merged, except that you need to sign the Google CLA (see [this comment](https://github.com/GoogleChrome/rendertron/pull/364#issuecomment-577467351)). Once you did and commented here, I'll land the filesystem cache :)

**Comment by chungwong** [comment]
> @chungwong Hey thanks again for your contribution! It's not ready to be merged, except that you need to sign the Google CLA (see [this comment](https://github.com/GoogleChrome/rendertron/pull/364#issuecomment-577467351)). Once you did and commented here, I'll land the filesystem cache :)

This is embarrassing, I have signed that earlier and I doubled checked it said I have already signed it.
I am not sure what went wrong actually.

**Comment by AVGP** [comment]
> > @chungwong Hey thanks again for your contribution! It's not ready to be merged, except that you need to sign the Google CLA (see [this comment](https://github.com/GoogleChrome/rendertron/pull/364#issuecomment-577467351)). Once you did and commented here, I'll land the filesystem cache :)
> 
> This is embarrassing, I have signed that earlier and I doubled checked it said I have already signed it.
> I am not sure what went wrong actually.

Don't worry, nothing embarrasing here :) Maybe you used a different email address in the commits you made? This might block the CLA to be recognized..

**Comment by chungwong** [comment]
> > > @chungwong Hey thanks again for your contribution! It's not ready to be merged, except that you need to sign the Google CLA (see [this comment](https://github.com/GoogleChrome/rendertron/pull/364#issuecomment-577467351)). Once you did and commented here, I'll land the filesystem cache :)
> > 
> > 
> > This is embarrassing, I have signed that earlier and I doubled checked it said I have already signed it.
> > I am not sure what went wrong actually.
> 
> Don't worry, nothing embarrasing here :) Maybe you used a different email address in the commits you made? This might block the CLA to be recognized..

Indded! I changed the email address in the commit and it worked

**Comment by AVGP** [comment]
Awesome, I'll get to this once I have a few minutes 👍 
Thanks for implementing this feature!

**Pull request merged**